### PR TITLE
3708 getblockhashesrpc

### PR DESCRIPTION
**Pull request by LarryRuane**
Part of #3708

### CODE REVIEW

**Comment by LarryRuane** [comment]
Please review commits on or after Apr 25 (the earlier commits are from other unmerged PRs).

**Comment by mms710** [comment]
Note for reviewers: these PRs must be reviewed and merged in this order:
#3863 
#3926 
#3953 
#3972 

Between every merge, ping Larry to rebase the next PR in the list so you can review and merge the next one down the line

**Comment by garethtdavies** [comment]
@LarryRuane running https://github.com/zcash/zcash/pull/3972/commits/160a3feaabd17472339727440b55e421a18cb38a but see `value error` for shielding tx such as 35d7fae0ef1a48c4370d8f5b2f295cfe1e725032136aa849554d4410267449f2

This is the current Insight explorer: https://explorer.zecmate.com/tx/35d7fae0ef1a48c4370d8f5b2f295cfe1e725032136aa849554d4410267449f2 and this is what is displayed running this branch:

insight-latest

**Comment by LarryRuane** [comment]
@garethtdavies, thank you for finding this, should be fixed in https://github.com/zcash/zcash/commit/55ff20540a81c9d7e2bdb80780bf4a735d8995c4

**Comment by garethtdavies** [comment]
Confirmed this fixes it running #55ff205. 

There seems to be a separate issue with fees which are all 0 (every single transaction).

fees

**Comment by LarryRuane** [comment]
@garethtdavies, I think it's fixed now (by commit 080fd80f1f329337814d35ea2009ac8ed4fe7ac4), please try latest commit, 8676a0f688a91319adb6739605d5243a3cff0fcd

Thanks so much for helping with the testing.

**Comment by LarryRuane** [comment]
@garethtdavies, update, sorry for the churn, but I don't think that will fix it. There was a key-value pair missing from the output of `getrawtransaction 1`, but upon closer inspection, that field is not displayed at all for transparent-to-transparent transactions, only for joinsplit transactions. I'm still investigating. 

One thing that would help is if you could send me the output of `getrawtransaction 1` for any transaction that is showing as having a zero fee (I think you said all of them do), and then I can compare that to the output for the same transaction as returned by the hackworks patched zcash. There must be some difference, because I can't imagine the fee is generated from anything else but the output of that RPC (it should be just the sum of the inputs minus the sum of the outputs).

**Comment by LarryRuane** [comment]
@garethtdavies, now I do believe the zero-fees bug is fixed (by 42a94842d2f7c87dfc2c486c2dc05affb5edf113). Please try that commit (or 7e9af4f6464b68713f23bf96d94b3456d3c392e1, the latest), thanks.

**Comment by garethtdavies** [comment]
Finally got around to testing this and confirm it is working great. Same block as above below with correct fees :+1:

**Comment by garethtdavies** [comment]
Running 56d7770b23d513a42a4c2bd12b1d1e26e103fe4a

There is an issue with unparsed addresses with those transactions in the mempool (by viewing them under "Latest Transactions". Here is an unconfirmed transaction with the unparsed address and incorrect fee.

mempool

As soon as this tx is confirmed it looks OK:

confirmed

This happens to all tx in the mempool e.g.

**Comment by LarryRuane** [comment]
Thanks, @garethtdavies, should be fixed (commit bfb5d0af9365e590a814fecdf99d867c94361b48 or later).

**Comment by garethtdavies** [comment]
Now running latest commit `v2.0.6-a2af276aa` and confirm this fixes it :smile: 

mempool-fix

File: qa/rpc-tests/timestampindex.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#
+# Test timestampindex generation and fetching
+#
+
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
```
Comment (Eirik0) [inline qa/rpc-tests/timestampindex.py]: I am not 100% sure, but I think this will fail the unused import check.
```suggestion
from test_framework.util import (
    assert_equal,
    connect_nodes,
    initialize_chain_clean,
    start_nodes,
    stop_nodes,
    wait_bitcoinds,
)
```

File: qa/rpc-tests/timestampindex.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python2
+# Copyright (c) 2014-2015 The Bitcoin Core developers
```
Comment (Eirik0) [inline qa/rpc-tests/timestampindex.py]: It looks like we updated the copyright in the other PRs

Comment (Eirik0) [reply qa/rpc-tests/timestampindex.py]: Also this should be `python` and not `python2`
```
#!/usr/bin/env python
# Copyright (c) 2019 The Zcash developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or https://www.opensource.org/licenses/mit-license.php .
```

File: src/rpc/misc.cpp [hunk 3]
```diff
@@ -1077,8 +1077,8 @@ UniValue getspentinfo(const UniValue& params, bool fHelp)
             "  ,...\n"
             "}\n"
             "\nExamples:\n"
-            + HelpExampleCli("getspentinfo", "'{\"txid\": \"0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9\", \"index\": 0}'")
-            + HelpExampleRpc("getspentinfo", "{\"txid\": \"0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9\", \"index\": 0}")
+            + HelpExampleCli("getspentinfo", "'{\"txid\": \"33990288fb116981260be1de10b8c764f997674545ab14f9240f00346333b780\", \"index\": 4}'")
+            + HelpExampleRpc("getspentinfo", "{\"txid\": \"33990288fb116981260be1de10b8c764f997674545ab14f9240f00346333b780\", \"index\": 4}")
```
Comment (str4d) [inline src/rpc/misc.cpp]: Why are these altered?

File: src/main.cpp [hunk 4]
```diff
@@ -4427,6 +4439,7 @@ bool static LoadBlockIndexDB()
     LogPrintf("%s: insight explorer %s\n", __func__, fAddressIndex ? "enabled" : "disabled");
```
Comment (str4d) [inline src/main.cpp]: There's a bug in this log line; it should be testing `fInsightExplorer`.

File: qa/rpc-tests/spentindex.py [hunk 5]
```diff
@@ -29,7 +32,8 @@ def setup_chain(self):
     def setup_network(self):
         # -insightexplorer causes spentindex to be enabled (fSpentIndex = true)
 
-        self.nodes = start_nodes(3, self.options.tmpdir,
+        self.nodes = start_nodes(
+            3, self.options.tmpdir,
```
Comment (str4d) [inline qa/rpc-tests/spentindex.py]: I'd prefer that this `getblockhashes` PR didn't include all the noise from formatting changes that, if desired, should have been in #3953.

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -126,94 +126,6 @@ UniValue blockheaderToJSON(const CBlockIndex* blockindex)
     return result;
 }
 
-// insightexplorer
-UniValue blockToDeltasJSON(const CBlock& block, const CBlockIndex* blockindex)
```
Comment (str4d) [inline src/rpc/blockchain.cpp]: We only just reviewed this as a separate function in #3953; please leave it where it was so this PR doesn't have the additional burden of reviewing moved code.

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -448,14 +360,14 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
...
-            "  \"previousblockhash\": \"hash\",\n"
-            "  \"nextblockhash\": \"hash\"\n"
+            "  \"time\" : n,                  (numeric) The block version\n"
+            "  \"mediantime\": n,             (numeric) The most recent blocks' ave time\n"
+            "  \"nonce\" : \"nonce\",           (hex string) The nonce\n"
+            "  \"bits\" : \"1d00ffff\",         (hex string) The bits\n"
+            "  \"difficulty\": n,             (numeric) the current difficulty\n"
+            "  \"chainwork\": \"xxxx\"          (hex string) total amount of work in active chain\n"
+            "  \"previousblockhash\" : \"hash\",(hex string) The hash of the previous block\n"
+            "  \"nextblockhash\" : \"hash\"     (hex string) The hash of the next block\n"
```
Comment (str4d) [inline src/rpc/blockchain.cpp]: Given how many of these cleanup changes there are, they should all be in a separate commit from the addition of `getblockhashes`.

File: src/rpc/client.cpp [hunk 8]
```diff
@@ -105,6 +105,10 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "getaddressdeltas", 0},
     { "getaddressutxos", 0},
     { "getaddressmempool", 0},
+    { "getblockhashes", 0},
+    { "getblockhashes", 1},
+    { "getblockhashes", 2},
+    { "getblockdeltas", 0},
```
Comment (str4d) [inline src/rpc/client.cpp]: The `getblockdeltas` change should have been in #3953. I'll have to go back and figure out why I missed it in that review.

File: src/rpc/misc.cpp (hunk 3 above)
Comment (LarryRuane) [reply src/rpc/misc.cpp]: That's so they match the documentation here: file:///home/larry/gitlab/zcash-docs/build/html/rtd_pages/insight_explorer.html (which hasn't been merged yet), and the reason for those values there is that all the examples actually work on testnet. (I think the previous value, `0437cd...` is from bitcoin.)

Comment (LarryRuane) [reply src/rpc/misc.cpp]: yes, it's from bitcoin mainnet block 9: https://blockchair.com/bitcoin/transaction/0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9

File: qa/rpc-tests/spentindex.py (hunk 5 above)
Comment (LarryRuane) [reply qa/rpc-tests/spentindex.py]: Okay, I was just trying to get pyflakes to complain less, but I'll undo those changes.

Comment (Eirik0) [reply qa/rpc-tests/spentindex.py]: Agreed, however, I don't mind if these formatting changes are included in this PR. It might have made more sense as a separate commit to distinguish it from the other parts of this PR, but I would be ok with including this as is.

File: src/rpc/client.cpp (hunk 8 above)
Comment (LarryRuane) [reply src/rpc/client.cpp]: Yes, sorry I missed that too; the reason I didn't catch it is because I added argument-processing tests to `src/test/rpc_tests.cpp` in the current PR; #3953 didn't include those tests for this RPC (I had missed it).

File: qa/rpc-tests/spentindex.py (hunk 5 above)
Comment (str4d) [reply qa/rpc-tests/spentindex.py]: Yep, I have no problem with the changes overall. The issue is that being so close to the RC, these kinds of changes just add noise to the PR and make it harder to review.

File: src/rpc/blockchain.cpp [hunk 9]
```diff
@@ -448,14 +449,14 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
...
-            "  \"time\": n,\n"
-            "  \"mediantime\": n,\n"
-            "  \"nonce\": \"hexstring\",\n"
-            "  \"bits\": \"hexstring\",\n"
-            "  \"difficulty\": ,\n"
-            "  \"chainwork\": \"hexstring\",\n"
-            "  \"previousblockhash\": \"hash\",\n"
-            "  \"nextblockhash\": \"hash\"\n"
+            "  \"time\" : n,                  (numeric) The block version\n"
+            "  \"mediantime\": n,             (numeric) The most recent blocks' ave time\n"
```
Comment (daira) [inline src/rpc/blockchain.cpp]:
```suggestion
            "  \"mediantime\": n,             (numeric) The 11 most recent blocks' median time\n"
```

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -485,6 +485,89 @@ UniValue getblockdeltas(const UniValue& params, bool fHelp)
...
+{
+    std::string enableArg = "insightexplorer";
+    bool fEnableGetBlockHashes = fExperimentalMode && fInsightExplorer;
+    std::string disabledMsg = "";
+    if (!fEnableGetBlockHashes) {
+        disabledMsg = experimentalDisabledHelpMsg("getblockhashes", enableArg);
+    }
+    if (fHelp || params.size() < 2)
+        throw runtime_error(
+            "getblockhashes high low ( {\"noOrphans\": true|false, \"logicalTimes\": true|false} )\n"
```
[the remark this answers is not in the document]
Comment (LarryRuane) [reply src/rpc/blockchain.cpp]: I thought the same thing, but ... compatibility! It will be nice if and when we control that insight nodejs code, and we can make the interfaces more sane, although we'd have to consider if there are other explorers out there that we still wouldn't control that will want to use our `zcashd`.

Also, I don't like the way the defaults work. IIRC, if either is zero, then the other is ignored, and the method returns the full range. It would be nice if `high` could be zero to mean from `low` to the current best height.

File: src/test/rpc_tests.cpp [hunk 11]
```diff
@@ -384,9 +385,17 @@ BOOST_AUTO_TEST_CASE(rpc_insightexplorer)
...
+    CheckRPCThrows("getblockhashes 0 0",
+        "Error: getblockhashes is disabled. "
+        "Run './zcash-cli help getblockhashes' for instructions on how to enable this feature.");
 
+    // During startup of the real system, fInsightExplorer ("-insightexplorer")
+    // automatically enables the next three, but not here, must explicitly enable.
     fExperimentalMode = true;
     fInsightExplorer = true;
+    fAddressIndex = true;
+    fSpentIndex = true;
```
[the remark this answers is not in the document]
Comment (LarryRuane) [reply src/test/rpc_tests.cpp]: What's mostly being checked in these tests are just argument processing (making sure JSON arguments are converted correctly, for example), and that happens before the enable checks. So the tests happen to all pass without these flags being enabled. But the tests for the latest one, `getblockhashes`, did need its corresponding enable flag to be set, and then I realized that the others should be set too.

File: src/rpc/blockchain.cpp (hunk 9 above)
Comment (LarryRuane) [reply src/rpc/blockchain.cpp]: I'll do this in that follow-up cleanup PR, since all three of you have already ACKed.

**Comment by daira** [comment]
reACK 70bda8416a410e2618e641b587d0bc845feec088

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Disk space failures on the workers. I cancelled the remaining builds and let the EC2 instances time out.

@zkbot retry

**Comment by str4d** [comment]
The transient `ld` failure on kitchensink strikes again. I'll retry in a bit.

**Comment by str4d** [comment]
@zkbot retry

**Comment by daira** [comment]
GitHub doesn't seem to recognize that this branch has been merged.
